syntax = "proto3";

package smart_devices;

// Mensagem genérica para um dispositivo inteligente
message SmartDevice {
  string device_id = 1;    // ID único do dispositivo
  string device_name = 2;  // Nome do dispositivo
  string device_type = 3;  // Tipo do dispositivo (e.g., Lamp, Sensor)
  bool is_on = 4;          // Estado ligado/desligado
  string device_ip = 5;    // IP do dispositivo
  int32 device_port = 6;   // Porta do dispositivo
}

// Mensagem para representar uma lâmpada
message Lamp {
  string device_id = 1;    // ID da lâmpada
  bool is_on = 2;          // Estado ligado/desligado
  float brightness = 3;    // Nível de brilho (0.0 - 1.0)
  string color = 4;        // Cor atual
}

// Mensagem para comandos enviados ao dispositivo
message DeviceCommand {
  string device_id = 1;      // ID do dispositivo alvo
  string command = 2;        // Comando enviado (e.g., "ligar", "desligar", "alterar cor")
  float brightness = 3;      // Nível de brilho (para lâmpadas)
  string color = 4;          // Cor desejada (para lâmpadas)
  map<string, string> parameters = 5; // Parâmetros adicionais para comandos personalizados
}

// Mensagem para resposta de comandos
message CommandResponse {
  bool success = 1;          // Indica se o comando foi bem-sucedido
  string message = 2;        // Mensagem de resposta (e.g., erro ou confirmação)
}

// Mensagem para descoberta de dispositivos
message DeviceDiscovery {
  string device_id = 1;      // ID do dispositivo
  string device_name = 2;    // Nome do dispositivo
  string device_type = 3;    // Tipo do dispositivo (e.g., Lamp, MotionSensor)
  bool is_on = 4;            // Estado ligado/desligado
  float brightness = 5;      // Brilho (somente para lâmpadas)
  string color = 6;          // Cor (somente para lâmpadas)
  string device_ip = 7;      // IP do dispositivo
  int32 device_port = 8;     // Porta do dispositivo
  bool motion_detected = 9;  // Indica se um movimento foi detectado (somente para sensores de movimento)
}

// Serviço para gerenciar dispositivos inteligentes
service SmartDeviceService {
  rpc SendCommand (DeviceCommand) returns (CommandResponse);
  rpc DiscoverDevices (DeviceDiscovery) returns (CommandResponse);
}
